### PR TITLE
Adds pytorch version of Hungarian algorithm

### PR DESCRIPTION
**Pull request by fakufaku**
This removes the necessity of brining back the loss to CPU during PIT.

### CODE REVIEW

**Comment by quancs** [comment]
Great job! I was doing this translation when implementing PIT, but I failed, because my translation works very slow.
So I use a [naive implementation](https://github.com/PyTorchLightning/metrics/blob/43a2afea4f7a8611b033d65358d97c580de08720/torchmetrics/functional/audio/pit.py#L172) when the speaker number is small. 
How is your speed on GPU compared with the scipy version? ^^

**Comment by fakufaku** [comment]
Hey @quancs ! The translation is quite slow compared to scipy '^_^
Scipy is very fast because it is implemented in C and not in python.
Nevertheless, using scipy means we need to bring back tensors on cpu for evaluation, which may cause performance issue especially for multi-GPU training.
Even the "slow" python version should not be that slow compared to actual separation/computation of metrics I think, although I have not carefully measured.
What is your experience there ?

I think the performance issue could be fixed by implementing this as a CPP extension via pytorch.

**Comment by quancs** [comment]
@fakufaku My translation on GPU seems to be slower than the scipy version.

I guess the slowness originated from the algorithm itself, as there are so many judgements, which is not the thing GPUs good at. So, to speed up, my idea is that we need an algorithm that could solve the assignment problem involving more vector and matrix operations.

Your idea should work, as the judgements are performed on python not cpp.
